### PR TITLE
ddl, util: Handle default values for generating column expressions when adding an index

### LINKED ISSUE #9311: Add an index to the virtual generated column whose expression contains the column added by "add column"
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a int);
insert into t values();
ALTER TABLE t ADD COLUMN d date DEFAULT '9999-12-31';
ALTER TABLE t ADD COLUMN d1 date as (DATE_SUB(d, INTERVAL 31 DAY));
ALTER TABLE t ADD INDEX idx_d(d1);
```

2. What did you expect to see?
Normal execution.

3. What did you see instead?
screen shot 2019-02-14 at 19 45 27

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v3.0.0-beta-58-g632a8c4c8
Git Commit Hash: 632a8c4c82f2c67dcd28b9a8fa10c543bc153fdb
Git Branch: master

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Fix #9311
Before this PR: 
```
tidb> create table t(id int primary key);
Query OK, 0 rows affected (0.01 sec)
tidb> insert into t values(1);
Query OK, 1 row affected (0.00 sec)
tidb> ALTER TABLE t ADD COLUMN d date DEFAULT '9999-12-31';
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD COLUMN d1 date as (DATE_SUB(d, INTERVAL 31 DAY));
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD INDEX idx_d(d1);
ERROR 1105 (HY000): cannot decode index value, because [types:1292]invalid time format: '{0 0 0 0 0 0 0}'
tidb> ALTER TABLE t ADD COLUMN id1 int as (id+1);
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD INDEX idx1(id1); // The idx1's value is 0 in store. 
Query OK, 0 rows affected (0.02 sec)
tidb> select id1 from t use index(idx1); // The idx'1 value is 0, but we will do `cast(plus(test.t.id, 1))`.
+------+
| id1  |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
```
Now:
```
tidb> create table t(id int primary key);
Query OK, 0 rows affected (0.01 sec)
tidb> insert into t values(1);
Query OK, 1 row affected (0.00 sec)
tidb> ALTER TABLE t ADD COLUMN d date DEFAULT '9999-12-31';
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD COLUMN d1 date as (DATE_SUB(d, INTERVAL 31 DAY));
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD INDEX idx_d(d1);
Query OK, 0 rows affected (0.07 sec)
tidb> ALTER TABLE t ADD COLUMN id1 int as (id+1);
Query OK, 0 rows affected (0.02 sec)
tidb> ALTER TABLE t ADD INDEX idx1(id1);
Query OK, 0 rows affected (0.02 sec)
tidb> select id1 from t use index(idx1);
+------+
| id1  |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
```

### What is changed and how it works?
Add default value and handle value to the row before evaluating the generated expression.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @winkyao @crazycs520

**Comment by zimulala** [comment]
PTAL @winkyao @crazycs520

File: util/rowDecoder/decoder.go [hunk 1]
```diff
@@ -79,8 +87,29 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, b []byte,
...
+		id := dCol.Info.ID
+		val, ok := row[id]
+		c, ok1 := rd.colsInGeneratedExpr[id]
+		if ok || !ok1 {
+			rd.mutRow.SetValue(rd.columns[id].Info.Offset, val.GetValue())
+			continue
+		}
+
+		// Get the default value of the column in the generated column expression.
+		if c.IsPKHandleColumn(rd.tbl.Meta()) {
```
Comment (crazycs520) [inline util/rowDecoder/decoder.go]: How about use `if rd.pkIsHandle && mysql.HasPriKeyFlag(dCol.Info.Flag)`, then we can only store `pkIsHandle` in RowDecoder, and remove `tbl`, `colsInGeneratedExpr`.

Comment (zimulala) [reply util/rowDecoder/decoder.go]: It seems that we can remove `tbl`, but why we can remove `colsInGeneratedExpr`？
I think storing `tbl` or `pkIsHandle` are both OK, and using`IsPKHandleColumn` is easier.

Comment (zimulala) [reply util/rowDecoder/decoder.go]: Removed `colsInGeneratedExpr`.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**